In [53]:
import polars as pl
import pandas as pd
import numpy as np
import bottleneck as bn


## Polars

Считаем датасет из файла `train.csv` (это данные о выживаемости на Титанике) с помощью polars

In [54]:
df = pl.read_csv('train.csv')

Выведем информацию согласно заданию

In [55]:
print(df)

print(df["Pclass"].value_counts()) #Посчитаем количество пассажиров каждого класса

print(df.group_by("Sex").agg((pl.col("Survived")).sum())) #Выведем количество выживших мужчин и женщин на корабле

print(df.filter(df["Age"] > 44)) #Выведем часть таблицы с пассажирами, возраст которых больше 44 лет

shape: (891, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 1           ┆ 0        ┆ 3      ┆ Braund, Mr. Owen ┆ … ┆ A/5 21171  ┆ 7.25    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Harris           ┆   ┆            ┆         ┆       ┆          │
│ 2           ┆ 1        ┆ 1      ┆ Cumings, Mrs.    ┆ … ┆ PC 17599   ┆ 71.2833 ┆ C85   ┆ C        │
│             ┆          ┆        ┆ John Bradley     ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Flor…           ┆   ┆            ┆   

## Ускорение работы с pandas

In [56]:
!pip install bottleneck


[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: C:\Users\anuta\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Считаем датасет из файла `train.csv` (это данные о выживаемости на Титанике) с помощью pandas

In [57]:
df = pd.read_csv('train.csv')

Выведем информацию согласно заданию

In [58]:
#Посчитайте средний возраст пассажиров и его стандартное отклонение
print(f'Средний возраст пассажиров\t {bn.nanmean(df.Age)}')
print(f'Стандартное отклонение возраста\t {bn.nanstd(df.Age)}')

#Для каждого пассажира умножьте значение столбца Fare на 1.3, если класс его билета - 1 или 2, и на 1.1, если класс его билета - 3. Сохраните результаты как новый столбец Fare_new. (При реализации расчета используйте любые альтернативы методу iterrows)
np_arr = df[['Fare', 'Pclass']].to_numpy()
new_col = [row[0]*(1.3*(row[1] < 3) + 1.1*(row[1] == 3)) for row in np_arr]
df.insert(10, 'Fare_new', new_col, True)
df

Средний возраст пассажиров	 29.69911764705882
Стандартное отклонение возраста	 14.516321150817317


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare  Fare_new Cabin Embarked  
0        0         A/5 21171   7.2500   7.97500   NaN        S  
1        0          PC 17599  71.2833  92.66829   C85        C  
2        0  STON/O2. 3101282   7.9250   8.71750   NaN        S  
3        0            113803  53.1000  69.03000  C123        S  
4        0            373450   8.0500   8.85500   NaN        S  
..     ...               ...      ...       ...   ...      ...  
886      0            211536  13.0000  16.90000   NaN        S  
887      0            112053  30.0000  39.00000   B42        S  
888      2        W./C. 6607  23.4500  25.79500   NaN        S  
889      0            111369  30.0000  39.00000  C148        C  
890      0            370376   7.7500   8.52500   NaN        Q  

[891 rows x 13 columns]

## Оптимизация типов pandas

Считайте датасет из файла `Housing.csv` (это данные о ценах домов) с помощью pandas

In [59]:
df = pd.read_csv('Housing.csv')

Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях

In [60]:
df

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [62]:
df.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

Определение оптимальных с точки зрения потребления памяти типов данных:

1. Строковые параметры для оптимизации следует категорировать.
2. Оставшиеся численные параметры не должны быть отрицательными, а значит для оптимизации нужно использовать unsigned типы.
3. Значения параметров bedrooms, bathrooms, stories, parking целочисленны и не выходят за границы [0;6], поэтому для них следует использовать uint8.
4. Значения параметра area также целочисленны и не выходят за границы [1650;16200], а значит для их хранения нужно использовать uint16.
5. Значения параметра price также целочисленны и не выходят за границы [1.75e+6;1.33e+7], а значит для их хранения ужно использовать uint32.

Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации

In [63]:
df['price'] = df['price'].astype('uint32')
df['area'] = df['area'].astype('uint16')
df['bedrooms'] = df['bedrooms'].astype('uint8')
df['bathrooms'] = df['bathrooms'].astype('uint8')
df['stories'] = df['stories'].astype('uint8')
df['parking'] = df['parking'].astype('uint8')
df['mainroad'] = df['mainroad'].astype('category')
df['guestroom'] = df['guestroom'].astype('category')
df['basement'] = df['basement'].astype('category')
df['hotwaterheating'] = df['hotwaterheating'].astype('category')
df['airconditioning'] = df['airconditioning'].astype('category')
df['prefarea'] = df['prefarea'].astype('category')
df['furnishingstatus'] = df['furnishingstatus'].astype('category')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    uint32  
 1   area              545 non-null    uint16  
 2   bedrooms          545 non-null    uint8   
 3   bathrooms         545 non-null    uint8   
 4   stories           545 non-null    uint8   
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    uint8   
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), uint16(1), uint32(1), uint8(4)
memory usage: 10.0 KB


В результате мы оптимизировали память больше чем в 5 раз.